File: AtCoder/ABC/B/267/B.cpp
#include<iostream>
#include<vector>
using namespace std;

bool is_standing(int index, vector<vector<int>> &pos){
    for(auto val: pos.at(index)){
        if(val == 1) return true;
    }
    return false;
}

int main(){
    // INPUT
    string s;
    cin >> s;   

    // MAKE POSITION
    vector<vector<int>> pos(7);
    int l=3, cnt=1, index=0;
    for(int i=0; i<4; i++){
        for(int j=0; j<cnt; j++){
            pos.at(l + 2*j).push_back(s.at(index) - '0');
            index++;
        }
        cnt++;
        l--;
    }
    /*
    // DISPLAY POS
    for(vector<int> ppos: pos){
        for(auto val: ppos){
            cout << val << ' ' ;
        }
        cout << endl;
    }
    */

    // COMBINATION
    bool is_split=false;
    for(int i=0; i<7; i++){
        if(s.at(0) == '1') break;
        for(int j=i+1; j<7; j++){
            if(is_standing(i, pos) && is_standing(j, pos)){
                for(int k=i+1; k<j; k++){
                    if(!is_standing(k, pos)) is_split = true;
                }
            }
        }
    }

    cout << (is_split ? "Yes" : "No") << endl;
}

File: AtCoder/ABC/B/267/B.rb
def is_standing(index, pos)
    pos[index].each do |val|
        return true if val == '1'
    end
    return false
end

s = gets.chomp
pos = Array.new(7){ Array.new }
index = 0
l = 3
cnt = 1
4.times do
    cnt.times do |i|
        pos[l + 2*i] << s[index]
        index += 1
    end
    l -= 1
    cnt += 1
end
is_split = false
[*0...7].combination(2).each do |c|
    break if s[0] == '1'
    l, r = c
    if is_standing(l, pos) and is_standing(r, pos)
        [*l..r].each do |mid|
            is_split = true if not is_standing(mid, pos)
        end
    end 
end
puts is_split ? "Yes" : "No"